Build CUDA under `host_injections` and make EESSI aware of host CUDA drivers

ocaisa opened the pull request.

ocaisa commented:
This can be tested by starting the build container from the PR and running, e.g., 
```
source /cvmfs/pilot.eessi-hpc.org/versions/2023.06/init/bash 
./gpu_support/nvidia/install_cuda_host_injections.sh 12.1.1
```

ocaisa commented:
But I need to transition the PR to (also?) use `eessi_container.sh`

ocaisa commented:
This PR can be tested without GPUs, use
```
./eessi_container.sh -m shell --nvidia install -v
```
to start the build container with the necessary settings, then you can install CUDA (under `host_injections`) with
```
source /cvmfs/pilot.eessi-hpc.org/versions/2023.06/init/bash 
./gpu_support/nvidia/install_cuda_host_injections.sh 12.1.1
```

@trz42 There are some sizable changes to the build container in the PR so you may wish to take it for a test drive.

casparvl commented on gpu_support/nvidia/install_cuda_host_injections.sh:
```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env bash
```
Should add some comment here on what this script is meant for. E.g.

```
This script can be used to install CUDA in the host_injections directory. This provides the parts of the CUDA installation that can not be redistributed as part of EESSI due to license limitations. While GPU-based software from EESSI will run without these, installation of additional CUDA software requires the CUDA installations in host_injections to be present.

The host_injections directory is a variant symlink that points to `/opt/eessi`, unless otherwise defined in the local CVMFS configuration (see https://cvmfs.readthedocs.io/en/stable/cpt-repo.html#variant-symlinks). For the installation to be succesful, this directory needs to be writeable to the user executing this script.
```

casparvl commented on scripts/utils.sh:
```diff
@@ -32,11 +32,81 @@ function check_exit_code {
+  return $return_code
+}
+
+function check_in_prefix_shell() {
```
Seems unused in this PR?

ocaisa replied:
I've left it in place as it is small and useful in later PRs

casparvl commented on scripts/utils.sh:
```diff
@@ -32,11 +32,81 @@ function check_exit_code {
+  fi
+}
+
+function float_greater_than() {
```
Seems unused in this PR?

casparvl commented on eessi_container.sh:
```diff
@@ -394,12 +432,30 @@ fi
+    if [[ "${NVIDIA_MODE}" == "install" || "${NVIDIA_MODE}" == "all" ]]; then
+        # Add additional bind mounts to allow CUDA to install within a container
+        EESSI_VAR_LOG=${EESSI_TMPDIR}/var-log
+        EESSI_USR_LOCAL_CUDA=${EESSI_TMPDIR}/usr-local-cuda
```
Why are these needed? Is this just because the installation of CUDA needs to paths to be available and/or writeable? maybe good to include in the comment above.

ocaisa replied:
The true answer is only something NVIDIA can now, these paths were decided based on trial and error

casparvl replied:
I figured you might have gotten a `permission denied` or something on these paths before, and that's why you decided to bind-mound them - if that's the case it'd be good to include that. But if you don't really know why they are needed, there's not much we can do :)

ocaisa replied:
I've added a clarification comment in https://github.com/EESSI/software-layer/pull/368/commits/090351864cbc5b97f35498b278d16041a5b97b1e

ocaisa commented on gpu_support/nvidia/install_cuda_host_injections.sh:
```diff
@@ -0,0 +1,92 @@
+
+  if [[ -z "${EBROOTEASYBUILD}" ]]; then
+    echo_yellow "Loading EasyBuild module to do actual install"
+    module load EasyBuild
```
I've added error catching in https://github.com/EESSI/software-layer/pull/368/commits/090351864cbc5b97f35498b278d16041a5b97b1e and made a few more updates to how we check for `eb`

ocaisa commented on gpu_support/nvidia/install_cuda_host_injections.sh:
```diff
@@ -0,0 +1,92 @@
+    module load EasyBuild
+  fi
+
+  # we need the --rebuild option and a (random) dir for the module since we are
```
Added in https://github.com/EESSI/software-layer/pull/368/commits/090351864cbc5b97f35498b278d16041a5b97b1e

ocaisa commented on gpu_support/nvidia/install_cuda_host_injections.sh:
```diff
@@ -0,0 +1,92 @@
+check_eessi_initialised
+
+if [[ $# -eq 0 ]] ; then
+    fatal_error "You must provide the CUDA version as an argument, e.g.:\n $0 11.3.1"
```
I've added changes in https://github.com/EESSI/software-layer/pull/368/commits/090351864cbc5b97f35498b278d16041a5b97b1e that should help people use the right version here (the one that mirrors what they have under EESSI) and does a good bit of related error checking

casparvl replied:
Lgtm

ocaisa commented on gpu_support/nvidia/install_cuda_host_injections.sh:
```diff
@@ -0,0 +1,92 @@
+  # We don't want hooks used in this install, we need a vanilla CUDA installation
+  touch "$tmpdir"/none.py
+  # shellcheck disable=SC2086  # Intended splitting of extra_args
+  eb --prefix="$tmpdir" ${extra_args} --hooks="$tmpdir"/none.py --installpath="${cuda_install_parent}"/ CUDA-"${install_cuda_version}".eb
```
I didn't try this either but the current approach isn't that dirty in my opinion (and I know it works!)

casparvl replied:
I tested it, it doesn't work with `--hooks=None`, so good thing you kept this :)

ocaisa commented on gpu_support/nvidia/install_cuda_host_injections.sh:
```diff
@@ -0,0 +1,198 @@
+  # We don't want hooks used in this install, we need a vanilla CUDA installation
+  touch "$tmpdir"/none.py
+  # shellcheck disable=SC2086  # Intended splitting of extra_args
+  eb --prefix="$tmpdir" ${extra_args} --hooks="$tmpdir"/none.py --installpath="${cuda_install_parent}"/ "${cuda_easyconfig}"
```
@boegel Why don't they have to accept the terms of the EULA here? I think we probably need to enforce that they explicitly do this

ocaisa replied:
Implemented in https://github.com/EESSI/software-layer/pull/368/commits/12719eccaabf57112332cc4a4958d0d03e382ef4

ocaisa commented:
@casparvl I've addressed your review as much as I can and I've added better command line options, including a critical acceptance of the CUDA EULA. This should be read for showtime now

ocaisa commented:
I think I'll fold the script to the linking to the host drivers in here as well...working on it now

casparvl commented on gpu_support/nvidia/install_cuda_host_injections.sh:
```diff
@@ -0,0 +1,213 @@
+  show_help
+  error="\nYou must provide a semantic version for CUDA (e.g., 12.1.1) via the appropriate\n"
+  error="${error}command line option. This script is intended for use with EESSI so the 'correct'\n"
+  error="${error}version to provide is probably the one that is available under\n"
```
Tiny detail: it might not be one. I guess it's a completely valid use case to have a CUDA installed in `host_injections` for _each_ cuda in `$EESSI_SOFTWARE_PATH/software/CUDA`. Not sure how I would phrase that here more clearly though... I'm also ok with leaving it like this.

ocaisa replied:
Rephrased in https://github.com/EESSI/software-layer/pull/368/commits/480d35643c7740c48f5508303e1cccadf04aae76

casparvl commented on gpu_support/nvidia/install_cuda_host_injections.sh:
```diff
@@ -0,0 +1,213 @@
+  error="\nYou must provide a semantic version for CUDA (e.g., 12.1.1) via the appropriate\n"
+  error="${error}command line option. This script is intended for use with EESSI so the 'correct'\n"
+  error="${error}version to provide is probably the one that is available under\n"
+  error="${error}$EESSI_SOFTWARE_PATH/software/CUDA\n"
```
Another small detail: you use `$EESSI_SOFTWARE_PATH` here in the error message, before checking it exists (that is currently on line 105 and 107). You could move up the check for `$EESSI_SOFTWARE_PATH` before this semantic-version check.

casparvl commented on gpu_support/nvidia/install_cuda_host_injections.sh:
```diff
@@ -0,0 +1,213 @@
+# Make sure EESSI is initialised
+check_eessi_initialised
+
+if [[ -z "${EESSI_SOFTWARE_PATH}" ]]; then
```
Isn't that what `check_eessi_initialised` already checks? I believe that even exits upon failure, so you wouldn't ever make it to this point if `EESSI_SOFTWARE_PATH`  wasn't defined, right?

ocaisa replied:
You're right, simplified in https://github.com/EESSI/software-layer/pull/368/commits/480d35643c7740c48f5508303e1cccadf04aae76

casparvl commented on gpu_support/nvidia/link_nvidia_host_libraries.sh:
```diff
@@ -0,0 +1,133 @@
+    found_paths+=("/sbin/$command_name")
+fi
+IFS=':' read -ra path_dirs <<< "$PATH"
+for dir in "${path_dirs[@]}"; do
```
As first part of this loop, do:
```
    if [ "$dir" = "/sbin" ]; then
        continue  # we've already checked for $command_name in /sbin, don't need to do it twice
    fi
```
Otherwise, if `/sbin` is also on `PATH`, `/sbin/ldconfig` gets added twice to `found_paths`:
```
Found ldconfig in the following locations:
- /sbin/ldconfig
- /sbin/ldconfig
```

ocaisa replied:
Adopted in https://github.com/EESSI/software-layer/pull/368/commits/480d35643c7740c48f5508303e1cccadf04aae76

casparvl commented on gpu_support/nvidia/link_nvidia_host_libraries.sh:
```diff
@@ -0,0 +1,133 @@
+  rm -r "$temp_dir"
+fi
+
+# Make latest symlink for NVIDIA drivers
```
I'm actually wondering: what is the point of the `latest` symlink here? You're doing a check that if the drivers don't match the linked drivers, you do it again. So there is only ever going to be _one_ copy of the driver symlinks here, right?

ocaisa replied:
Not once we support compatibility libraries

ocaisa replied:
In that scenario we would have a script that figures out which version of the CUDA compat libraries that we ship work with the driver and then have `latest` point to that

casparvl replied:
Ah right, I didn't think about the compat libs here. Makes sense

casparvl commented on gpu_support/nvidia/link_nvidia_host_libraries.sh:
```diff
@@ -0,0 +1,133 @@
+  target_path=$(readlink -f "$host_injection_linker_dir/lib")
+  if [ "$target_path" != "$$host_injections_nvidia_dir/latest" ]; then
+    cd $host_injection_linker_dir
+    ln -sf $host_injections_nvidia_dir/latest lib
```
I'm wondering if this is the way to go. It'll create a link like:
```
/cvmfs/pilot.eessi-hpc.org/host_injections/2023.06/compat/linux/x86_64/lib -> /cvmfs/pilot.eessi-hpc.org/host_injections/nvidia/x86_64/latest
```
But what if we want to inject other libraries to be picked up later? If `/cvmfs/pilot.eessi-hpc.org/host_injections/2023.06/compat/linux/x86_64/lib` is a symlink, it can only point one way.

It brings me to the broader question: why do we even install this in `/cvmfs/pilot.eessi-hpc.org/host_injections/nvidia/x86_64/latest` first? Why not directly create symlinks in `/cvmfs/pilot.eessi-hpc.org/host_injections/2023.06/compat/linux/x86_64/lib`? You probably had a good reason for it I assume, but it's not immediately clear to me :)

ocaisa replied:
You're right, this is not the way to go, but it is what is supported right now by the EESSI linker. What I would like to do in the next compat layer is have a few additional trusted directories:
* a generic one as is right now
* one for nvidia
* one for AMD

It's better not to put symlinks in the trusted dir directly so that when we have a global update of the drivers, we only have to make a single update to the symlinks (rather than for every version of EESSI)
